File: week1/pset/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

// For Check if the number is valid
int digits_not_multiplied(long);
int digits_multiplied(long);
int digits_counter(long);

bool isvalid(long);

int main(void)
{
    // Prompt for Credit Number
    long input = get_long("Number: ");

    // Counting Digits
    int digits = digits_counter(input),
        valid  = isvalid(input);

    // This is for checking
    input /= pow(10, (digits - 2));

    // VISA Check
    if ((digits == 13 || digits == 16) && input / 10 == 4 && valid)
    {
        printf("VISA\n");
    }
    // AMEX Check
    else if (digits == 15 && (input == 34 || input == 37) && valid)
    {
        printf("AMEX\n");
    }
    // MASTER CARD Check
    else if (valid && digits == 16 && (input == 51 || input == 52 || input == 53 || input == 54 || input == 55))
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }


    return 0;
}

int digits_not_multiplied(long i)
{
    // Calculate The sum of not multiplied digits
    int sum = 0;

    while (i != 0)
    {
        sum += i % 10;
        i /= 100;
    }
    return sum;
}

int digits_multiplied(long i)
{
    // Calculate The sum of multiplied digits
    i /= 10;
    int sum = 0;

    while (i != 0)
    {
        int temp = (i % 10) * 2;
        if (temp < 10)
        {
            sum += temp;
        }
        else
        {
            sum += (temp % 10);
            temp /= 10;
            sum += temp;
        }
        i /= 100;
    }
    return sum;
}

int digits_counter(long i)
{
    // Calculate digits of number
    int digits = 0;

    while (i != 0)
    {
        digits++;
        i /= 10;
    }
    return digits;
}

bool isvalid(long i)
{
    // Check if the number is valid or not
    int sum_M = digits_multiplied(i),
        sum_NotM = digits_not_multiplied(i),
        temp;
    temp = sum_M + sum_NotM;
    if (temp % 10 == 0)
    {
        return 1;
    }
    return 0;
}

File: week6/pset/sentimental-credit/credit.py
# Define main function
def main():
    # Promt the user for input
    card = get_card()
    # Counting Digits
    digits = digits_count(card)
    valid = isvalid(card)
    # This is for checking
    card //= pow(10, (digits - 2))
    if (digits == 13 or digits == 16) and card // 10 == 4 and valid:
        print("VISA")
    # AMEX Check
    elif digits == 15 and (card == 34 or card == 37) and valid:
        print("AMEX")
    # MASTER CARD Check
    elif (
        valid
        and digits == 16
        and (card == 51 or card == 52 or card == 53 or card == 54 or card == 55)
    ):
        print("MASTERCARD")
    else:
        print("INVALID")


def get_card():
    while True:
        try:
            n = int(input("Number: "))
            if n > 0:
                return n
        except ValueError:
            print("Not a number")


# Calculate number digits
def digits_count(n):
    digits = 0
    while n != 0:
        digits += 1
        n //= 10
    return digits


# Calculate The sum of multiplied digits
def sum_digits_multiplied(n):
    n //= 10
    sum = 0
    while n != 0:
        temp = (n % 10) * 2
        if temp < 10:
            sum += temp
        else:
            sum += temp % 10
            temp //= 10
            sum += temp
        n //= 100
    return sum


# Calculate The sum of not multiplied digits
def sum_digits_not_multiplied(n):
    sum = 0
    while n != 0:
        sum += n % 10
        n //= 100
    return sum


# Check if the number is valid or not
def isvalid(n):
    sum_M = sum_digits_multiplied(n)
    sum_NotM = sum_digits_not_multiplied(n)
    temp = sum_M + sum_NotM
    if temp % 10 == 0:
        return True
    else:
        return False


# Calling main
main()
